sematic/plugins/abstract_metrics_storage.py

```diff
@@ -0,0 +1,204 @@
+        One of `MetricType`. Using the str value instead of Enum instance for
+        serializability.
+    series: List[Tuple[float, Tuple[str, ...]]]
+        The metric series. A list of tuples. Every element in the list
+        correspond to a series value. The tuple contains two elements: the
+        value, and a tuple of label values. The order of label values
+        corresponds to the order of label names in group_by_labels.
+    group_by_labels: List[str]
+        The list of label names the metric was aggregated over.
+
+    Examples
+    --------
+    Run count with no aggregations.
+    ```
+    MetricSeries(
+        metric_name="sematic.func_run_count",
+        metric_type="COUNT",
+        series=[(483.0, ())],
+        group_by_labels=[]
+    )
+    ```
+
+    Success rate by calculator path by day
+    ```
+    MetricSeries(
+        metric_name="sematic.func_success_rate",
+        metric_type="GAUGE",
+        series=[
+            (0.68, ("path.to.foo", "2023-04-12")),
+            (0.95, ("path.to.foo", "2023-04-11")),
+            (0.45, ("path.to.bar", "2023-04-12")),
+            ...
+        ],
+        group_by_labels=["calculator_path", "date"],
+    )
+    ```
+    """
+
+    metric_name: str
+    metric_type: Optional[str] = None
+    series: List[Tuple[float, Tuple[str, ...]]] = field(default_factory=list)
+    group_by_labels: List[str] = field(default_factory=list)
+
+
+class GroupBy(enum.Enum):
+    """
+    Options to aggregate metrics over.
+
+    `GroupBy.timestamp` will aggregate over the value's timestamp, which is
+    likely unique. Therefore, this aggregation option will return every single
+    value/
+    """
+
+    timestamp = "timestamp"
+    date = "date"
+    run_id = "run_id"
+    calculator_path = "calculator_path"
+    root_id = "root_id"
+    root_calculator_path = "root_calculator_path"
+
+
+class NoMetricError(Exception):
+    def __init__(self, metric_name: str, plugin_name: str):
+        super().__init__(f"No metric named {metric_name} was found in {plugin_name}")
+
+
+class AbstractMetricsStorage(abc.ABC):
+    """
+    Abstract base class for metrics storage plug-ins.
+    """
+
+    @abc.abstractmethod
+    def store_metrics(self, metric_points: Sequence[MetricPoint]) -> None:
+        """
+        Persists a list of metric points.
+
+        Parameters
+        ----------
+        metric_points: Iterable[MetricPoint]
+            List of metric points to persist.
+        """
+        pass
+
+    @abc.abstractmethod
+    def get_metrics(self, labels: MetricsLabels) -> Tuple[str, ...]:
+        """
+        Gets the list of unique metric names for the given set of labels.
+
+        Parameters
+        ----------
+        labels: MetricsLabels
+            Labels by which to filter metric names.
+        """
+        pass
+
+    @abc.abstractmethod
+    def get_aggregated_metrics(
+        self, filter: MetricsFilter, group_by: Sequence[GroupBy]
+    ) -> MetricSeries:
```

Comment:
Imagine something like a loss curve, where a new value is added 10 times a second. The user wants to view this metric over the course of the last 10 hours. We don't want to give them back every single point, and thus overwhelm the front-end. But aggregating by date is too broad. How will the aggregation/bucketing be done in such a case? This goes back to the idea that time should be unique, and not just another label to be grouped by. Datadog, for example, won't actually give you the raw metrics values but will *always* aggregate over time and put things in time-based buckets (of Datadog's own choosing, I believe).

Reply:
See previous reply about treating date as a label (TL;DR it is not).

A loss curve is typically not reported 10 times a second, but more typically once per epoch.
In this initial version, I was thinking of not yet supporting other aggregation than the ones in `GroupBy`.

In a future iteration, we can add a `GroupBy.minute` or so.

Reply:
Though you're right that I misunderstood treating date as a label, my point still stands about third-party metrics storage solutions not naturally lend themselves to using "group by" semantics for time-based fields. See [these docs](https://docs.datadoghq.com/dashboards/functions/rollup/#rollup-interval-enforced-vs-custom) about intervals between data points from Datadog.

The default behavior is for you to not even specify the interval between points and they automatically decide on one. If you specify one, they may override it based on the size of the overall time window you're querying over. If we try to treat date/hour/minute as something you can "group by" then it will be really hard to integrate with such solutions. I think an approach that's more sustainable is to basically just *do* what Datadog does if you don't explicitly `rollup`: we let people give a "rollup_hint" parameter for the time which can be some interval specified in seconds. We try to rollup such that points are at the desired interval, but don't guarantee that we can do it. This captures the fact that the time-based "grouping" may not be exact, though other groupings will be.

Reply:
We still need a way to differentiate "I want a single scalar for the entire time window" from "I want a time series across the time window". Right now, the former is the default when no `GroupBy.date` or `GroupBy.timestamp` is passed.
What we could do is
- `rollup_hint=None` – no time series, aggregate all into a single number
- `rollup_hint=N` – attempt to build a time series in intervals of N seconds, if N is greater than a hard-coded threshold similar to the ones on the Datadog docs page.